classes/query/Repository.php

```diff
@@ -0,0 +1,218 @@
+use PKP\mail\Mailable;
+use PKP\note\Note;
+use PKP\notification\Notification;
+use PKP\notification\NotificationSubscriptionSettingsDAO;
+use PKP\security\Role;
+use PKP\stageAssignment\StageAssignment;
+use PKP\user\User;
+
+class Repository
+{
+    /**
+     * Fetch a query by symbolic info, building it if needed.
+     */
+    public function build(int $assocType, int $assocId, int $userId, int $stageId, float $seq = 0, bool $closed = false): Query
+    {
+        return Query::withUserId($userId)
+            ->withAssoc($assocType, $assocId)
+            ->firstOr(fn() => Query::create([
+                'assocType' => $assocType,
+                'assocId' => $assocId,
+                'userId' => $userId,
+                'stageId' => $stageId,
+                'seq' => $seq,
+                'closed' => $closed
+            ]));
+    }
+
+    /**
+     * Retrieve a count of all open queries totalled by stage
+     *
+     * @param int[] $participantIds Only include queries with these participants
+     *
+     * @return array<int,int> [int $stageId => int $count]
+     */
+    public function countOpenPerStage(int $submissionId, ?array $participantIds = null)
+    {
+        $counts = DB::table('queries as q')
+            ->when($participantIds !== null, function ($q) use ($participantIds) {
+                $q->join('query_participants as qp', 'q.query_id', '=', 'qp.query_id')
+                    ->whereIn('qp.user_id', $participantIds);
+            })
+            ->where('q.assoc_type', '=', Application::ASSOC_TYPE_SUBMISSION)
+            ->where('q.assoc_id', '=', $submissionId)
+            ->where('q.closed', '=', 0)
+            ->select(['q.stage_id', DB::raw('COUNT(q.stage_id) as count')])
+            ->groupBy(['q.stage_id'])
+            ->get()
+            ->mapWithKeys(fn ($row, $key) => [$row->stage_id => $row->count])
+            ->toArray();
+
+        return collect(Application::get()->getApplicationStages())
+            ->mapWithKeys(fn ($stageId, $key) => [$stageId => $counts[$stageId] ?? 0])
+            ->toArray();
+    }
+
+    /**
+     * Sequentially renumber queries in their sequence order.
+     *
+     * @param int $assocType Application::ASSOC_TYPE_...
+     * @param int $assocId Assoc ID per assocType
+     */
+    public function resequence($assocType, $assocId): void
+    {
+        $result = Query::withAssoc($assocType, $assocId)
+            ->orderBy('seq')
+            ->get();
+
+        $result->each(function ($item, $key = 1) {
+            Query::where('query_id', $item->queryId)
+                ->update(['seq' => $key]);
+        });
+    }
+
+    /**
+     * Start a query
+     *
+     * Inserts the query, assigns participants, and creates the head note
+     *
+     * @return int The new query id
+     */
+    public function addQuery(int $submissionId, int $stageId, string $title, string $content, User $fromUser, array $participantUserIds, int $contextId, bool $sendEmail = true): int
```

Comment:
In progress - is only used in this file (in `addCommentsforEditorsQuery` below).